io/jvm/src/main/scala/scalaz/effect/RTS.scala

```diff
@@ -471,22 +455,16 @@ private object RTS {
                       }
                     } else {
                       // Error caught:
-                      // FIXME this cast is not safe, `err` is defined as `E => IO[E2, B]`
-                      val handledIo = errorHandler(error).asInstanceOf[IO[E, Any]]
-
-                      if (finalizer eq null) {
-                        // No finalizer to run:
-                        curIo = handledIo
+                      curIo = nextInstr[E](error, stack)
```

Comment:
This should be correct: the error handler was just pushed onto the stack, and now it's being used to compute the next `IO` using the thrown `error`.

Reply:
By the way, this is the point where the stack is used and the error handler is popped off the stack to compute the next IO based on the error.